File: python/2_simple_linear_regression.py
import pandas
from matplotlib import pyplot
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

dataset = pandas.read_csv('resources/Salary_Data.csv')

x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

# train the model
regression = LinearRegression()
regression.fit(X_train, y_train)

# save the function y <- x
y_predicted = regression.predict(X_train)

# plot the training set against the function
pyplot.scatter(X_train, y_train, marker='.', color='red')
pyplot.plot(X_train, y_predicted)
pyplot.xlabel('Experience (years)')
pyplot.ylabel('Salary')
pyplot.show()

# compare the real test outcomes against the predicted (function)
# how well our model predicted the outcomes
pyplot.scatter(X_test, y_test, marker='.', color='red')
pyplot.plot(X_train, y_predicted)
pyplot.xlabel('Experience (years)')
pyplot.ylabel('Salary')
pyplot.show()

# use the model to predict salary of employee with 12 years of experience
prediction = regression.predict([[12]])
print(f'employee with 12 years of experience should get {prediction}')

# equation is <code>y = coef_ * x + intercept_</code>
print(f'y = {regression.coef_} * x + {regression.intercept_}')

File: R/2_simple_linear_regression.R
# Title     : Simple linear regression
# Created by: Oleksandr_Gavryliuk
# Created on: 13-Jul-21

dataset <- read.csv('..\\resources\\Salary_Data.csv')

library(caTools)
set.seed(123)
split <- sample.split(dataset$Salary, SplitRatio = 2/3)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)
print(training_set)
print(test_set)

# fit the linear model
regressor <- lm(formula = Salary ~ YearsExperience, data = training_set)
summary(regressor)

# predict salaries for the test set
y_predicted <- predict(regressor, test_set)

# predict salary for employee with 12 years of experience
twelve_years_salary <- predict(regressor, data.frame("YearsExperience" = 12))

# plot
library(ggplot2)
ggplot() +
  geom_point(aes(training_set$YearsExperience, training_set$Salary),
             colour = 'red') +
  geom_line(aes(training_set$YearsExperience, predict(regressor, training_set)),
            colour = 'blue') +
  ggtitle('Salary vs Experience') +
  xlab('Experience (years') +
  ylab('Salary (US dollars')

ggplot() +
  geom_point(aes(test_set$YearsExperience, test_set$Salary),
             colour = 'red') +
  geom_line(aes(training_set$YearsExperience, predict(regressor, training_set)),
            colour = 'blue') +
  ggtitle('Salary vs Experience (test set)') +
  xlab('Experience (years)') +
  ylab('Salary (US dollars)')